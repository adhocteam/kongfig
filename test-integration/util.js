import adminApi from '../src/adminApi';
import readKongApi from '../src/readKongApi';
import execute from '../src/core';
import { logReducer } from '../src/kongStateLocal';
import getCurrentStateSelector from '../src/stateSelector';
import invariant from 'invariant';
import pad from 'pad';
import { pretty } from '../src/prettyConfig';

invariant(process.env.TEST_INTEGRATION_KONG_HOST, `
    Please set ${'TEST_INTEGRATION_KONG_HOST'.bold} env variable

    TEST_INTEGRATION_KONG_HOST=localhost:8001 yarn test

    ${'WARNING! Running integration tests are going to remove all data from the kong'.red.bold}.
`);

const UUIDRegex = /[a-f0-9]{8}-?[a-f0-9]{4}-?4[a-f0-9]{3}-?[89ab][a-f0-9]{3}-?[a-f0-9]{12}/g;
const IGNORED_KEYS = ['created_at', 'version', 'orderlist', 'updated_at']
let uuids = {};
let log = [];
let rawLog = [];

export const exportToYaml = pretty('yaml');
export const getLocalState = () => getCurrentStateSelector(rawLog.reduce(logReducer, undefined));

export const testAdminApi = adminApi({
    host: process.env.TEST_INTEGRATION_KONG_HOST,
    https: false,
    ignoreConsumers: false,
    cache: false,
});

export const getLog = () => log;
export const logger = message => {
    if (message.type === 'experimental-features') {
        // cannot include these in tests because they change based on test matrix
        return;
    }

    const m = cloneObject(message);

    if (m.hasOwnProperty('uri')) {
        m.uri = m.uri.replace(process.env.TEST_INTEGRATION_KONG_HOST, 'localhost:8001');
    }

    rawLog.push(m);
    log.push(ignoreKeys(m));
};

const _ignoreKeys = (obj) => {
    if (typeof obj !== 'object') {
        return obj;
    }

    return obj && Object.keys(obj).reduce((x, key) => {
        if (typeof obj[key] === 'string' && obj[key].match(UUIDRegex)) {
            const value = obj[key].match(UUIDRegex).reduce((value, uuid) => {
                if (!uuids.hasOwnProperty(uuid)) {
                    const id = pad(12, `${Object.keys(uuids).length + 1}`, '0');
                    uuids[uuid] = `2b47ba9b-761a-492d-9a0c-${id}`;
                }

                return value.replace(uuid, uuids[uuid]);
            }, obj[key]);

            return addElement(x, value, key);
        } else if (IGNORED_KEYS.includes(key)) {
            return addElement(x, `___${key}___`, key);
        }

        return addElement(x, _ignoreKeys(obj[key]), key);
    }, Array.isArray(obj) ? [] : {});
};

const addElement = (obj, elem, key) => {
    if (Array.isArray(obj)) {
        return [ ...obj, elem ];
    } else {
        return { ...obj, [key]: elem }
    }
}

const cloneObject = obj => JSON.parse(JSON.stringify(obj));

export const ignoreKeys = (obj) => _ignoreKeys(cloneObject(obj));

const cleanupKong = async () => {
    const results = await readKongApi(testAdminApi);
    await execute({
        apis: results.apis.map(api => ({ ...api, ensure: 'removed' })),
        consumers: results.consumers.map(consumer => ({ ...consumer, ensure: 'removed' })),
        plugins: results.plugins.map(plugin => ({ ...plugin, ensure: 'removed' })),
        upstreams: results.upstreams.map(upstream => ({ ...upstream, ensure: 'removed' })),
        certificates: results.certificates.map(certificate => ({ ...certificate, ensure: 'removed' })),
        services: results.services.map(service => ({ ...markRoutesRemoved(service), ensure: 'removed' })),
    }, testAdminApi);
};

const markRoutesRemoved = (service) => {
    return { ...service, routes: service.routes.map((route) => ({ ...route, ensure: 'removed' })) };
}

export const tearDown = async () => {
    uuids = {};
    log = [];
    rawLog = [];
    await cleanupKong();
};
