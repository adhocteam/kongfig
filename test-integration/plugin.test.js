import execute from '../src/core';
import { testAdminApi, logger, getLog, tearDown } from '../test-integration/util';
import 'core-js/features/object/from-entries';

beforeEach(tearDown);
jest.setTimeout(10000);

describe('global plugins', () => {
  it('should properly swap username for consumer id', async () => {
    const config = {
      plugins: [
        {
          name: 'rate-limiting',
          attributes: {
            enabled: true,
            username: 'user',
            config: {
              second: 5,
              redis_database: 0,
              policy: 'cluster',
              hide_client_headers: false,
              redis_timeout: 2000,
              redis_port: 6379,
              limit_by: 'consumer',
              fault_tolerant: true
            }
          }
        }
      ],
      consumers: [
        {
          username: 'user',
          ensure: 'present'
        }
      ]
    }

    await execute(config, testAdminApi, logger);
    await execute(config, testAdminApi, logger);

    expect(getLog()).toMatchSnapshot();
  });
});
