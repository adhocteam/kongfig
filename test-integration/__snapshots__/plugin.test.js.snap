// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`global plugins should properly swap username for consumer id 1`] = `
Array [
  Object {
    "type": "kong-info",
    "version": "___version___",
  },
  Object {
    "params": Object {
      "body": Object {
        "username": "user",
      },
      "endpoint": Object {
        "name": "consumers",
      },
      "method": "POST",
      "type": "create-customer",
    },
    "type": "request",
    "uri": "http://localhost:8001/consumers",
  },
  Object {
    "content": Object {
      "created_at": "___created_at___",
      "custom_id": null,
      "id": "2b47ba9b-761a-492d-9a0c-000000000001",
      "tags": null,
      "username": "user",
    },
    "ok": true,
    "params": Object {
      "body": Object {
        "username": "user",
      },
      "endpoint": Object {
        "name": "consumers",
      },
      "method": "POST",
      "type": "create-customer",
    },
    "status": 201,
    "statusText": "Created",
    "type": "response",
    "uri": "http://localhost:8001/consumers",
  },
  Object {
    "params": Object {
      "body": Object {
        "config": Object {
          "fault_tolerant": true,
          "hide_client_headers": false,
          "limit_by": "consumer",
          "policy": "cluster",
          "redis_database": 0,
          "redis_port": 6379,
          "redis_timeout": 2000,
          "second": 5,
        },
        "consumer": Object {
          "id": "2b47ba9b-761a-492d-9a0c-000000000001",
        },
        "enabled": true,
        "name": "rate-limiting",
      },
      "endpoint": Object {
        "name": "plugins",
        "params": Object {
          "pluginName": "rate-limiting",
        },
      },
      "method": "POST",
      "type": "add-global-plugin",
    },
    "type": "request",
    "uri": "http://localhost:8001/plugins",
  },
  Object {
    "content": Object {
      "config": Object {
        "day": null,
        "fault_tolerant": true,
        "hide_client_headers": false,
        "hour": null,
        "limit_by": "consumer",
        "minute": null,
        "month": null,
        "policy": "cluster",
        "redis_database": 0,
        "redis_host": null,
        "redis_password": null,
        "redis_port": 6379,
        "redis_timeout": 2000,
        "second": 5,
        "year": null,
      },
      "consumer": Object {
        "id": "2b47ba9b-761a-492d-9a0c-000000000001",
      },
      "created_at": "___created_at___",
      "enabled": true,
      "id": "2b47ba9b-761a-492d-9a0c-000000000002",
      "name": "rate-limiting",
      "protocols": Array [
        "http",
        "https",
      ],
      "route": null,
      "run_on": "first",
      "service": null,
      "tags": null,
    },
    "ok": true,
    "params": Object {
      "body": Object {
        "config": Object {
          "fault_tolerant": true,
          "hide_client_headers": false,
          "limit_by": "consumer",
          "policy": "cluster",
          "redis_database": 0,
          "redis_port": 6379,
          "redis_timeout": 2000,
          "second": 5,
        },
        "consumer": Object {
          "id": "2b47ba9b-761a-492d-9a0c-000000000001",
        },
        "enabled": true,
        "name": "rate-limiting",
      },
      "endpoint": Object {
        "name": "plugins",
        "params": Object {
          "pluginName": "rate-limiting",
        },
      },
      "method": "POST",
      "type": "add-global-plugin",
    },
    "status": 201,
    "statusText": "Created",
    "type": "response",
    "uri": "http://localhost:8001/plugins",
  },
  Object {
    "type": "kong-info",
    "version": "___version___",
  },
  Object {
    "params": Object {
      "consumer": Object {
        "ensure": "present",
        "username": "user",
      },
      "noop": true,
      "type": "noop-consumer",
    },
    "type": "noop",
  },
  Object {
    "params": Object {
      "noop": true,
      "plugin": Object {
        "attributes": Object {
          "config": Object {
            "fault_tolerant": true,
            "hide_client_headers": false,
            "limit_by": "consumer",
            "policy": "cluster",
            "redis_database": 0,
            "redis_port": 6379,
            "redis_timeout": 2000,
            "second": 5,
          },
          "consumer": Object {
            "id": "2b47ba9b-761a-492d-9a0c-000000000001",
          },
          "enabled": true,
        },
        "name": "rate-limiting",
      },
      "type": "noop-global-plugin",
    },
    "type": "noop",
  },
]
`;
